added the searchbar, filters, and sort by dropdown box

0Jsnipes opened the pull request:
## Description

Provide a concise summary of the changes in this PR. Include the motivation behind these changes and any relevant context. 
I added the search bar, filter buttons, and sort by drop-down box.  these were asked to be added in the #196 issue 
I also added the favorites section and recommended tools section but did not implement them yet I was unsure of the functionality of the favorites and just started working with another collaborator on the issue. 

## Related Issue
#196 
If this PR addresses an issue, link to it here.
https://github.com/marvelai-org/marvel-platform/issues/196

## Type of Change

Please select the type(s) of change that apply and delete those that do not. 


- [x] **New feature**: A non-breaking change that adds functionality.


## Proposed Solution

Describe your code changes in detail. Explain how you implemented your solution and any design decisions you made. 

## How to Test

 Provide instructions on how to test these changes. Include details on test configurations, test cases, and expected outcomes. 

## Unit Tests

 List the unit tests added or modified to verify your changes. 
added 3 tests to ensure the functionality of the features 
filter.test.jsx
searchbar.test.jsx
sortdropdown.jsx

## Documentation Updates

Indicate whether documentation needs to be updated due to this PR. 

- [ ] Yes
- [ ] No

If yes, describe what documentation updates are needed and link to the relevant documentation.

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

## Additional Information

 Add any other information that might be useful for the reviewers. 
![Screenshot 2025-01-03 173227](https://github.com/user-attachments/assets/f869b1cd-b9b2-479d-a929-69fa42c482ff)

0Jsnipes committed:
made the pres generator need to style

0Jsnipes committed:
presentation generator deleted from my personal

0Jsnipes committed:
got search bar and filters added need to add sort feature and styling

0Jsnipes committed:
made search bar and sorting buttons

0Jsnipes committed:
sort works now

0Jsnipes committed:
made the search bar and filter buttons and sort by buttons and organized them correctly

0Jsnipes committed:
added reccomended and favorites sections

0Jsnipes committed:
sort,filter,search are done

0Jsnipes committed:
the filter search and sort are done filter does not currently filter anything but removes the tools when filtering

0Jsnipes committed:
added tests

0Jsnipes committed:
got the star to show and you can click on it without it breaking but it makes a star off to the side for some reason

Cosmodocus commented on templates/HomePage/HomePage.jsx:
Please remove 'React' from the import statement. Latest React versions no longer require this

Cosmodocus commented on templates/HomePage/HomePage.jsx:
```diff
@@ -1,93 +1,130 @@
+  const [currentTab, setCurrentTab] = useState('All');
+  const [searchQuery, setSearchQuery] = useState('');
+  const [sortOption, setSortOption] = useState('Most Popular');
+  const [favorites, setFavorites] = useState([]); // State to track favorite tool IDs
```
Lets also remove these unnecessary comments

Cosmodocus commented on templates/HomePage/HomePage.jsx:
```diff
@@ -1,93 +1,130 @@
+  const handleToggleFavorite = (id) => {
+    setFavorites((prevFavorites) =>
+      prevFavorites.includes(id)
+        ? prevFavorites.filter((favId) => favId !== id) // Remove if already in favorites
```
Remove these comments

Cosmodocus commented on templates/HomePage/HomePage.jsx:
```diff
@@ -1,93 +1,130 @@
+  // Filters and Search
+  const renderFilters = () => (
+    <Grid container direction="column" spacing={2}>
+      <Grid
```
Some of these components should have their styling attributes abstracted over onto the styles.js page. Lets emphasize seperation of concerns and follow what other files are doing as well.

Cosmodocus commented on templates/HomePage/HomePage.jsx:
```diff
@@ -1,93 +1,130 @@
-              key={tab}
-            />
-          ))}
+  // Filters and Search
```
Remove this comment

Cosmodocus commented on tools/components/Filter/Filters.jsx:
```diff
@@ -0,0 +1,36 @@
+import { Box, Grid } from '@mui/material';
+
+// tabs is an array of strings, activeTab is a string, setActiveTab is a function they are listing on the homepage
```
So if you want to leave descriptions for components, it would be best if you places a JSDocs comment with the data types of the components props, as well as its expected typing, rather than just a comment in itself.

This way you'll have left meaningful comments about the file,  as well as type safety, killing 2 birds with 1 stone

Cosmodocus commented on tools/components/Filter/Filters.jsx:
```diff
@@ -0,0 +1,36 @@
+      {tabs.map((tab) => (
+        <Grid item key={tab} xs>
+          <Box
+            component="button"
```
Abstract these styling attributes into their own styles.js folder please

Cosmodocus commented on tools/components/Filter/Filters.jsx:
```diff
@@ -0,0 +1,36 @@
+import { Box, Grid } from '@mui/material';
```
I understand why you placed this component here, however, this component could possibly be reused for other features. I think it would be wise to move these into the components folder in root directory. Ensure they all have their own index.js file, as well as their own styles.js file. Also go ahead and give each of these component a JSDocs comment with their own brief and meaningful descriptions of the components

Cosmodocus commented on tools/components/SearchBar/SearchBar.jsx:
Same advice for Filters as for here

Cosmodocus commented on tools/components/SortDorpdown/SortDropdown.jsx:
Here as well, move this to the root directory components folder with the same advice as the other components prior to this.

Cosmodocus commented on tools/components/ToolCard/ToolCard.jsx:
```diff
@@ -18,9 +18,7 @@ import { TOOLS_ID } from '@/tools/libs/constants/tools';
-const ToolCard = (props) => {
-  const { id, maskedToolUrl, backgroundImgURL, name, description } = props;
-
+const ToolCard = ({ id, name, description, maskedToolUrl, backgroundImgURL, favorites = [], handleToggleFavorite }) => {
```
move new and existing props into its own line for better readability. Have the argument of this component as 'props' as before, but underneath, have a line of code that follows:
`const ToolCard = (props) => {`
          `const { id, name, description, etc...} = props`

Cosmodocus commented on tools/components/ToolCard/ToolCard.jsx:
```diff
@@ -59,7 +57,22 @@ const ToolCard = (props) => {
+            </Typography>
+          <IconButton
+            onClick={(e) => {
+              e.stopPropagation(); // Prevent the click from propagating to the card
```
remove this comment

Cosmodocus commented on tools/views/ReccomendedForYou/ReccomendedForYou.jsx:
Let's ensure we keep spelling in check as well. Rename this to RecommendedForYou.jsx, as well as the folder.

Cosmodocus commented on tools/views/ReccomendedForYou/ReccomendedForYou.jsx:
```diff
@@ -0,0 +1,51 @@
+  );
+};
+
+ReccomendedForYou.propTypes = {
```
This is not necessary. It would be better to use JSDocs for this case.

Cosmodocus commented on tools/views/ToolsListingContainer/ToolsListingContainer.jsx:
```diff
@@ -12,29 +12,34 @@ const DEFAULT_TOOLS = new Array(8)
-      </Grid>
-    );
-  };
+const ToolsListingContainer = ({
```
Rework all the props to be destructured into its own variable called props, similar to how it was before, but with the new props you've added as well.

Cosmodocus reviewed (changes requested):
Hey @0Jsnipes, left some comments down below. Please take a look when you get a chance and ask any questions that you may have.

Also would be great if you could provide a short video of you going through the features.

0Jsnipes committed:
favorites section renders tool card but not fully yet

0Jsnipes committed:
renders half the card but the star does not fill with color

0Jsnipes committed:
super close it i need the favorites container to have the star and to disappear when there are none

Know167 committed:
added user favorite tools functionality

updated userSlice.js for adding pending, fulfilled and rejected slices for the updateUserFavorite function.
updated user.js by adding updateUserFavorite Async function writing to firestore.
updated HomePage.jsx to dispatch updateUserFavorite action, and update local 'favorites' state variable, and useEffect to update current state if user-state in redux store changes.

0Jsnipes committed:
Merge pull request #1 from Know167/develop

added user favorite tools functionality

0Jsnipes committed:
Merge branch 'develop' into develop
